discovery_client.go

```diff
@@ -0,0 +1,148 @@
+package rendezvous
+
+import (
+	"context"
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+type rendezvousDiscoveryClient struct {
+	rp        RendezvousPoint
+	peerCache sync.Map //is a map[string]discoveredPeerCache
+	rng       *rand.Rand
+	rngMux    sync.Mutex
+}
+
+type discoveredPeerCache struct {
+	cachedRegs map[peer.ID]*Registration
+	cookie     []byte
+	mux        sync.Mutex
+}
+
+func NewRendezvousDiscoveryClient(host host.Host, rendezvousPeer peer.ID) discovery.Discovery {
+	rp := NewRendezvousPoint(host, rendezvousPeer)
+	return &rendezvousDiscoveryClient{rp, sync.Map{}, rand.New(rand.NewSource(rand.Int63())), sync.Mutex{}}
+}
+
+func (c *rendezvousDiscoveryClient) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return 0, err
+	}
+
+	ttl := options.Ttl
+	var ttlSeconds int
+
+	if ttl == 0 {
+		ttlSeconds = 7200
+	} else {
+		ttlSeconds = int(math.Round(ttl.Seconds()))
+	}
+
+	if returnedTTL, err := c.rp.Register(ctx, ns, ttlSeconds); err != nil {
+		return 0, err
+	} else {
+		return returnedTTL, nil
+	}
+}
+
+func (c *rendezvousDiscoveryClient) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	const maxLimit = 1000
+	limit := options.Limit
+	if limit == 0 || limit > maxLimit {
+		limit = maxLimit
+	}
+
+	// Get cached peers
+	var cache *discoveredPeerCache
+
+	genericCache, _ := c.peerCache.LoadOrStore(ns, &discoveredPeerCache{})
+	cache = genericCache.(*discoveredPeerCache)
+
+	cache.mux.Lock()
+
+	// Remove all expired entries from cache
+	currentTime := int(time.Now().Unix())
+	newCacheSize := len(cache.cachedRegs)
+
+	for p := range cache.cachedRegs {
+		reg := cache.cachedRegs[p]
+		if reg.Ttl < currentTime {
+			newCacheSize--
+			delete(cache.cachedRegs, p)
+		}
+	}
+
+	cookie := cache.cookie
+	cache.mux.Unlock()
+
+	// Discover new records if we don't have enough
+	var discoveryErr error
+	if newCacheSize < limit {
+		if discoveryRecords, newCookie, err := c.rp.Discover(ctx, ns, limit, cookie); err == nil {
+			cache.mux.Lock()
+			if cache.cachedRegs == nil {
+				cache.cachedRegs = make(map[peer.ID]*Registration)
+			}
+			for i := range discoveryRecords {
```

Comment:
`for i, rec := range direcoveryRecords { ... }`

Reply:
That change actually breaks the below code unless you have a new variable for the modified rec because `Discover` returns `[]Registration` and not `[]*Registration`.

Modifying `rec` creates a new `Registration` that needs a variable anyway, so why not just call `rec` the pointer to the `Registration` we need?

Reply:
maybe we need a better algorithm, but ok.

Reply:
I think this is generally a problem with arrays of structs (instead of pointers to structs). Given that you'd like a new expiry struct anyway I can now iterate over `i, rec` however I'm pretty sure it'll result in an extra copy of the fields in `rec`. Any preferences on if I should change it to `i, rec`?

Reply:
you can't mutate this way.

Reply:
well we won't be mutating anyway since we need a new struct => copy

Reply:
struct copy is cheap, don't worry about it. So yeah, do it with a proper range loop.